Commit graph doc (#2044)
dolthub/docs: content/architecture/storage-engine/commit-graph.md

---
title: Commit Graph
---

Dolt's unique [storage engine](https://docs.dolthub.com/architecture/storage-engine) implements a Git-style commit graph of [Prolly Trees](https://docs.dolthub.com/architecture/storage-engine/prolly-tree). Dolt's commit graph facilitates common version control operations like log, diff, branch and merge on database tables instead of files.

![Dolt commit graph](../../.gitbook/assets/commit-graph-featured.png)

> # Git vs Dolt

Reviewer: Small thing, but technically the title is h1 and these should be h2

A commit contains two sets of information: the content and the metadata.

In Git, the content is the set of files as they existed at that point in time, identified by a content address. In Dolt, the content is the set of tables in the database at that point in time, identified by a content address. In Dolt, content addresses are created using a novel data structure called a [Prolly Tree](https://docs.dolthub.com/architecture/storage-engine/prolly-tree), that allows for structural sharing, efficient diff, and fast querying of table data.

Additionally, commit metadata like author, date, and message are stored so it is easier to identify the commit you are looking for in the version history. This metadata is considered when creating the content address that you see in the commit log. So, even if two commits have the exact same content but are committed at different times or by different authors, they will have different commit hashes. 

# Why put Commits in a Graph?

> Putting Commits is a graph allows for a representation of history, branches, and merges; core concepts of version control. A branch allows for multiple evolving histories. A merge allows two disparate histories to be combined.

Reviewer: ```suggestion
Putting commits in a graph allows for a representation of history, branches, and merges; core concepts of version control. A branch allows for multiple evolving histories. A merge allows two disparate histories to be combined.
```

`WORKING` is often called the working set. An interesting way to think about the working set is traditional file systems that don't use Git only have a working set. Traditional databases like MySQL or Postgres only have a working set. If you create a database in Dolt and only run traditional SQL, your working set will look and act exactly like a MySQL database.

## History

Commits are created using the aptly named `commit` command. When you commit your `STAGED` changes, the content that is staged is moved to the tip of the branch and you have an opportunity to add metadata like message and author. The `HEAD` of the branch `main` becomes this newly created commit. 

![A Second Commit](../../.gitbook/assets/commit-graph-second-commit.png)

Commits have zero to many parents. The init commit has zero parents. A normal commit has one parent, representing the previous commit metadata and content. A merge commit, which we'll discuss later, has many parents. 

Parents allow for the history of branches to be computed by walking the branch from its `HEAD`. This is commonly called the commit log and generated using the `log` command.

> 

Reviewer: Could be worth giving the example output of running `log` on `main` at this point. Like "Using the log command on the 'main' branch here would list commits `h512kl` and `t1ms3n`."

## Branches

Up to this point, we are dealing only with linear history. If there is only one editor making serial changes, the commit graph will look like a long line of commits. A linear commit graph is still a graph, but not a very interesting graph. 

> Branches allow for non-linear history, a fork in the commit graph. Branches are often used to isolate multiple user's changes. Two users can make changes to content without worrying about what the other is changing. This capability is quite powerful. We've all worked on a shared document where people stomp on each others changes. Branches prevent stomping.

Reviewer: ```suggestion
Branches allow for non-linear history, a fork in the commit graph. Branches are often used to isolate multiple users' changes. Two users can make changes to content without worrying about what the other is changing. This capability is quite powerful. We've all worked on a shared document where people stomp on each other's changes. Branches prevent stomping.
```

Branches are created using the `branch` command. When branches are created the `HEAD` of the branch points at a specified commit, usually the `HEAD` commit of the branch you are currently using.

![New branch](../../.gitbook/assets/commit-graph-new-branch.png)

> Now, using the same process above we can make a a commit on the branch. The `HEAD` of the new branch now points at this new commit. 

Reviewer: ```suggestion
Now, using the same process above we can make a commit on the branch. The `HEAD` of the new branch now points at this new commit. 
```
